modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/StringUtils.java

```diff
@@ -31,7 +34,11 @@
     private static Cache<Pair<String, Boolean>, String> camelizedWordsCache;
 
     // A cache of underscored words, used to optimize the performance of the underscore() method.
-    private static Cache<String, String> underscoreWords;
+    private static Cache<String, String> underscoreWordsCache;
+
+    // A cache of escaped words, used to optimize the performance of the escape() method.
+    private static Cache<EscapedNameOptions, String> escapedWordsCache;
```

Comment:
I'm looking at the method again and I'm trying to think of scenarios where there would be conflicts across threads if all the parameters are the same, and I can't think of any. I think you're probably fine with your current implementation for memoizing the escape function. I thought I had noticed something when I commented yesterday that prevented this function from being pure, and I'm not seeing it today.

But because you asked about a tying to the generation, if we ever wanted to have per-run caching we could do something like the following (for something like a CodegenModel cache for `$ref`!):

We could create a context object with unique ID as a hash key (an entity in DDD-land), and use this as the Cache key or as a component of the type defining the cache key.

For instance:

## Id

```
package org.openapitools.codegen.context;

import java.util.Objects;
import java.util.UUID;

public final class Id {
    private final UUID value;

    public Id() {
        this.value = UUID.randomUUID();
    }

    @Override
    public int hashCode() {
        return Objects.hash(getValue());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Id id = (Id) o;
        return getValue().equals(id.getValue());
    }

    public UUID getValue() {
        return value;
    }
}
```

## InvocationContext

```
package org.openapitools.codegen.context;

import java.util.Objects;
import java.util.UUID;

public final class InvocationContext {
    private final Id id;

    public InvocationContext() {
        this.id = new Id();
    }

    public UUID getIdentifier() {
        return this.id.getValue();
    }

    @Override
    public int hashCode() {
        // InvocationContext is an entity and should only maintain id in hashcode.
        return Objects.hash(id);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        InvocationContext that = (InvocationContext) o;

        // InvocationContext is an entity and should only maintain id in equals.
        return id.equals(that.id);
    }
}
```

And then update GlobalSettings to provide this per-run specific identifiable object:

```
public class GlobalSettings {
    private static ThreadLocal<InvocationContext> context = new InheritableThreadLocal<InvocationContext>() {
        @Override
        protected InvocationContext initialValue() {
            return new InvocationContext();
        }
    };
    private static ThreadLocal<Properties> properties = new InheritableThreadLocal<Properties>() {
        @Override
        protected Properties initialValue() {
            return (Properties) System.getProperties().clone();
        }
    };

    public static String getProperty(String key, String defaultValue) {
        return properties.get().getProperty(key, defaultValue);
    }

    public static String getProperty(String key) {
        return properties.get().getProperty(key);
    }

    public static void setProperty(String key, String value) {
        properties.get().setProperty(key, value);
    }

    public static void clearProperty(String key) {
        properties.get().remove(key);
    }

    public static InvocationContext getContext() {
        return context.get();
    }

    public static void reset() {
        properties.remove();
        context.remove();
    }
}
```

Then, you'd just need to work out how you'd want to work with this in your cache. For your escape function, for example, you could change `EscapedNameOptions` to accept a UUID which is added to equals and hashcode, and then construct it with the invocation context's identifier. So,

```
EscapedNameOptions ns = new EscapedNameOptions(
    GlobalSettings.getContext().getIdentifier(),
    name, replacementMap.keySet(), charactersToAllow, appendToReplacement);
```

in Thread1 would be different than the same in Thread2.

I could see us using this for a really simple ref lookup like this as the key and CodegenModel as the value:

```
package org.openapitools.codegen.context;

import java.util.Objects;
import java.util.UUID;

public final class ContextualRef {
    private final UUID invocationId;
    private final String ref;

    public ContextualRef(UUID invocationId, String ref) {
        if (ref == null) throw new IllegalArgumentException("ref can't be null");
        this.invocationId = invocationId;
        // normalize/validate ref?
        this.ref = ref;
    }

    @Override
    public int hashCode() {
        return Objects.hash(invocationId, ref);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ContextualRef that = (ContextualRef) o;
        return invocationId.equals(that.invocationId) &&
                ref.equals(that.ref);
    }
}
```